### PR TITLE
[NETWORK] Add Task to Respond to Requests from Peers

### LINKED ISSUE #2649: [LIBP2P] - Provide VID Shares When Requested
### What is this task and why do we need to work on it?

After #2648 the leader should be getting requests for VID shares from some replicas, some of the time.  This means the leader can remove VID share sending from the libp2p path and only respond to these requests.  The goal of this task is to have the mechanism lookup the share and respond in kind.

### What work will need to be done to complete this task?

- [ ] Create a Lookup Task which has a channel to the libp2p networking task
- [ ] Lookup task should have access to consensus storage, specifically the VID storage
- [ ] Libp2p task forwads the VID Request, and response Channel to the Lookup task via the channel
- [ ] Lookup tasks finds the VID task in storage and sends it on the `ResponseChannel`
- [ ] If no VID share is found, send `NotFound` instead

### Are there any other details to include?

See the [RequestResponse docs](https://docs.rs/libp2p/latest/libp2p/request_response/index.html), specifically [ResponseChannel](https://docs.rs/libp2p/latest/libp2p/request_response/struct.ResponseChannel.html) and [send_response](https://docs.rs/libp2p/latest/libp2p/request_response/struct.Behaviour.html#method.send_response)

Note: We might need to have a channel from the lookup task -> libp2p network task to send the info for a response

### What are the acceptance criteria to close this issue?

Libp2p works standalone without the CDN and without sending VID shares from the leader, except when requested

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by bfish713**
Closes #2649 
Depends on: #2679 

### This PR: 

This PR hooks up the changes from #2679 to the consensus layer, but only for handling requests, not sending them.

- Adds VID shares to the shared consensus state
- Creates a new task for handling requests
- The task:
  - Validates the Request by checking the sender has stake, and has signed the message
  - Looks up the VID shares for the view requested.
  - If the shares exist, looks up the senders share in the map
  - if it exists sends a VIDDisperse to the requester.  
  - IF at any point there is a failure we send NotFound to the requester.

A couple things to note:

- Only 1 request is handled at a time, and if there channel to handle requests becomes full, we start to drop the requests.  This is to protect us from getting overwhelmed.  
- The VID disperse information we send to the requester will still have the leaders signature on it but we strip out the shares not related to the requester.  The signature is over the VID commitment so the receiver can still verify
- The signature is over the hash of the serialized request, we can probably do better than that but I don't feel there is any attack this way.  
- We don't filter for old view numbers because that should naturally happen because we only keep undecided views in our state.

### This PR does not: 

This doesn't actually spawn the task, or implement the requesting side of things.

### Key places to review: 

The new task
Adding VID shares to consensus

### CODE REVIEW

File: crates/task-impls/src/response.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        &self,
+        mut vid: Proposal<TYPES, VidDisperse<TYPES>>,
+        key: TYPES::SignatureKey,
+    ) -> Message<TYPES> {
+        let Some(share) = vid.data.shares.get(&key) else {
+            return self.make_msg(ResponseMessage::NotFound);
+        };
+        let mut single_share_map = BTreeMap::new();
+        single_share_map.insert(key, share.clone());
+        vid.data.shares = single_share_map;
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/response.rs:136]: Yeah was looking for something like that

File: crates/task-impls/src/response.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                }
+            }
+        }
+    }
+
+    /// Handle an incoming message.  First validates the sender, then handles the contained request.
+    /// Sends the response via `chan`
+    async fn handle_message(&self, req: Message<TYPES>, chan: ResponseChannel<Message<TYPES>>) {
+        let sender = req.sender.clone();
+        if !self.valid_sender(&sender) {
```
[the remark this answers is not in the document]
Comment (bfish713) [reply crates/task-impls/src/response.rs:88]: Yeah I meant to add a Denied to the DataRequest enum, I will update with that

File: crates/task-impls/src/response.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+/// Type alias for the channel that we receive requests from the network on.
+type ReqestReceiver<TYPES> = mpsc::Receiver<(Message<TYPES>, ResponseChannel<Message<TYPES>>)>;
+
+/// Task state for the Network Request Task. The task is responsible for handling
+/// requests sent to this node by the network.  It will validate the sender,
+/// parse the request, and try to find the data request in the consensus stores.
+pub struct NetworkRequestState<TYPES: NodeType> {
+    /// Locked consensus state
+    consensus: LockedConsensusState<TYPES>,
```
Comment (jparr721) [inline crates/task-impls/src/response.rs:40]: Does this allow us to basically dispatch to consensus when we receive certain responses? How come we don't just broadcast in this case?

**Review by jparr721 — COMMENTED** [review]
One quick question.

File: crates/task-impls/src/response.rs (hunk 3 above)
Comment (bfish713) [reply crates/task-impls/src/response.rs:40]: It's the opposite, it allows us to read from the consensus state and respond to whoever is requesting the data.  The requests here are coming from external parties over the network layer.

**Review by rob-maron — APPROVED** [review]

**Pull request merged**